rmf_fleet_adapter/src/rmf_fleet_adapter/events/PerformAction.cpp

```diff
@@ -0,0 +1,248 @@
+PerformAction::Standby::Standby(
+  rmf_task_sequence::events::PerformAction::Description description)
+: _description(std::move(description))
+{
+  // Do nothing
+}
+
+//==============================================================================
+auto PerformAction::Standby::state() const -> ConstStatePtr
+{
+  return _state;
+}
+
+//==============================================================================
+rmf_traffic::Duration PerformAction::Standby::duration_estimate() const
+{
+  return _time_estimate;
+}
+
+//==============================================================================
+auto PerformAction::Standby::begin(
+  std::function<void()>,
+  std::function<void()> finished) -> ActivePtr
+{
+  if (!_active)
+  {
+    _active = Active::make(
+      _assign_id,
+      _context,
+      _description.category(),
+      _description.description(),
+      _time_estimate,
+      _state,
+      _update,
+      std::move(finished));
+  }
+
+  return _active;
+}
+
+//==============================================================================
+auto PerformAction::Active::make(
+  const AssignIDPtr& id,
+  agv::RobotContextPtr context,
+  const std::string& category,
+  nlohmann::json desc,
+  rmf_traffic::Duration time_estimate,
+  rmf_task::events::SimpleEventStatePtr state,
+  std::function<void()> update,
+  std::function<void()> finished) -> std::shared_ptr<Active>
+{
+  auto active = std::make_shared<Active>(
+    Active(std::move(category), std::move(desc), time_estimate));
+  active->_assign_id = id;
+  active->_context = std::move(context);
+  active->_update = std::move(update);
+  active->_finished = std::move(finished);
+  active->_state = std::move(state);
+  active->_execute_action();
+  active->_expected_finish_time =
+    active->_context->now() + time_estimate;
+  return active;
+}
+
+//==============================================================================
+PerformAction::Active::Active(
+  const std::string& category,
+  nlohmann::json desc,
+  rmf_traffic::Duration time_estimate)
+: _action_category(std::move(category)),
+  _action_description(std::move(desc)),
+  _time_estimate(std::move(time_estimate))
+{
+  // Do nothing
+}
+
+//==============================================================================
+auto PerformAction::Active::state() const -> ConstStatePtr
+{
+  return _state;
+}
+
+//==============================================================================
+rmf_traffic::Duration PerformAction::Active::remaining_time_estimate() const
+{
+  if (_context->action_remaining_time().has_value())
+  {
+    return _context->action_remaining_time().value();
+  }
+
+  // If an estimate is not provided we compute one based on the expected finish
+  // time
+  return _expected_finish_time - _context->now();
+}
+
+//==============================================================================
+auto PerformAction::Active::backup() const -> Backup
+{
+  // PerformAction doesn't need to be backed up
```

Comment:
In the future we could consider adding a function to `ActionExecution` that allows the system integrator to periodically send in backup messages. But for now, I think this is fine.

Reply:
Noted. Updated the comment.